Source/Core/Core/State.cpp

```diff
@@ -342,7 +342,7 @@ static void CompressAndDumpState(CompressAndDumpState_args save_args)
 
   // Setting up the header
   StateHeader header;
-  strncpy(header.gameID, SConfig::GetInstance().GetGameID().c_str(), 6);
+  memcpy(header.gameID, SConfig::GetInstance().GetGameID().c_str(), 6);
```

Comment:
Again, this is copying into a string buffer, using the `str` variant of copying is correct here.

Reply:
std::array<char, 6> gameID;  // The Game ID
This is not a string. GameID is not null terminate string at all.

Reply:
`gameID` isn't a std::array. `std::string` also provides copying to buffers as part of its interface:

```cpp
SConfig::GetInstance().GetGameID().copy(header.gameID, 6);
```

Reply:
Check this: SConfig is a different class here.

struct StateHeader
{
  char gameID[6];
  u32 size;
  double time;
};

 // Setting up the header
  StateHeader header;
  memcpy(header.gameID, SConfig::GetInstance().GetGameID().c_str(), 6);

Reply:
I don't particularly see how that's relevant. Regardless of what it is (copying characters into a character buffer), this is directly [provided](https://en.cppreference.com/w/cpp/string/basic_string/copy) by `std::string`, so using `memcpy` or `strnlen` isn't even necessary here.

Reply:
are u confused by the src and dest?
char *strncpy(char *dest, const char *src, size_t n);

Reply:
`strncpy` would be correct here. We cannot use `memcpy`, because there is a possibility that the memory pointed to by `SConfig::GetInstance().GetGameID().c_str()` is less than 6 bytes, making us read invalid memory. As for `std::string::copy`, doesn't this function not write a null terminator to the destination? That is something we would want in case the game ID is shorter than 6 bytes. (Or has the entire `gameID` array been set to null bytes at some prior point?)

Reply:
GameID is not a STRING. It is just handled as a string in the SConfig. When we convert the string back to its native format, we should do some special handling here. Strncpy doesn't make sense as we are copy string to a non-string buffer.

Maybe I should introduce a new type gameid_t to abstract this out.

Reply:
`gameID` is a string buffer.

Reply:
(lldb) p header->gameID
(char [6]) $2 = "SBLE5G"

You can find the gameID is a char[6] that doesn't have space for '\0'. Do you think this is a string still?

Reply:
It is a string buffer which contains strings that are not necessarily null terminated. There's no rule that says that bytes in memory representing a string have to end with a null pointer – it is a convention and not a requirement. Though of course, if the string is not null terminated, you have to be careful with what functions you pass the string to. In our case, we only pass it to `strncmp` while making sure to specify the size as 6, so it's fine.

But regardless of whether you consider `gameID` to contain a string, we clearly cannot use `memcpy` for the reason I described in my first comment, otherwise there is a risk of undefined behavior.

Reply:
make sense. Let me think how to fix this.

Reply:
I quite literally gave you a solution ten hours ago. You can even dehardcode the 6 in the process.

```cpp
SConfig::GetInstance().GetGameID().copy(header.gameID, std::size(header.gameID));
```

Reply:
Just make sure to also change `StateHeader header;` to `StateHeader header{};` if you use that solution.